### PR TITLE
Feature/refactoring

### PR DESCRIPTION
**Pull request by mbauer55**

### CODE REVIEW

File: buildRule.js [hunk 1]
```diff
@@ -204,434 +154,29 @@ const buildCondition = (inputObject) => {
...
 
-  } else {
-    const updatedVariableName = variableNames.join(',');
-    const updatedObject = object.replaceAll('$', '')
+  const ruleValue = buildRulePart(topLevelMap, ruleName);
 
-    object = `(${updatedVariableName})${functionCall}${updatedObject}`;
-    fact = `(${updatedVariableName})${functionCall}${objectName}(${updatedVariableName}).${factValue}`;
-    condition = `(${updatedVariableName})${functionCall}${factName}(${updatedVariableName}) ${operatorString} ${targetString}`;
+  console.log(ruleValue)
```
Comment (madhavKD) [inline buildRule.js:165]: Remove log, if not required

File: buildRule.js [hunk 2]
```diff
@@ -204,434 +154,29 @@ const buildCondition = (inputObject) => {
...
-}
+  // objectMap.forEach((value, key) =>{
+  //   console.log(createFunction(key, value))
+  // })
+  // factMap.forEach((value, key) =>{
+  //   console.log(createFunction(key, value))
+  // })
+  // ruleMap.forEach((value, key) =>{
+  //   console.log(createFunction(key, value))
+  // })
```
Comment (madhavKD) [inline buildRule.js:175]: Remove theses commented lines, if not required

File: buildRule.js [hunk 3]
```diff
@@ -204,434 +154,29 @@ const buildCondition = (inputObject) => {
...
 }
 
-// buildRule(requestBody)
-build(requestBodyFact)
+build(requestBody);
+// build(requestBodyFact)
 // build(requestBodyFact2)
-build(requestBodyCondition)
-
+// build(requestBodyCondition)
```
Comment (madhavKD) [inline buildRule.js:182]: Remove these commented code, if not required

File: buildRule.js [hunk 4]
```diff
@@ -173,9 +126,6 @@ const build = (obj) => {
   // console.log(flatt)
```
Comment (madhavKD) [inline buildRule.js:126]: Remove this

File: buildParts.js [hunk 5]
```diff
@@ -0,0 +1,369 @@
+const { convertGraphQLToFQL } = require("./graphqlToFQLConverter");
```
Comment (firehudson) [inline buildParts.js:1]: Format file with indentation of 2 spacing

File: buildParts.js [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+	for (let index = 0; index < data.length; index++) {
+		const { all, any } = data[index];
+		if (all)
+			getQueries(all, queries);
+		else if (any)
+			getQueries(any, queries);
+		else {
+			const { source, target } = data[index];
+			if (source?.type?.toLowerCase() === 'fact')
```
Comment (firehudson) [inline buildParts.js:25]: Store `'fact'` in constant variable like this `const FACT = 'fact'` and use variable everywhere in the code

File: buildParts.js [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+				queries.push(source.value);
+			if (target?.type?.toLowerCase() === 'fact')
+				queries.push(target.value);
+		}
+	}
+}
+
+const getCorrectOperator = (operator) => {
```
Comment (firehudson) [inline buildParts.js:33]: Convert this function in single line arrow function

File: buildParts.js [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+	return operatorMap[operator.toLowerCase()] || operator;
+}
+
+const getTargetString = (operatorString, target) => {
+	const { type, value } = target;
+
+	if (type.toLowerCase() === 'fact')
```
Comment (firehudson) [inline buildParts.js:40]: Replace `'fact'` with constant variable

File: buildParts.js [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+			? `${convertGraphQLToFQL(value)}`
+			: `(${convertGraphQLToFQL(value)})`;
+	else if (type.toLowerCase() === 'string')
+		return operatorString === '==' ? `"${value}"` : `("${value}")`
+
+	return value;
+}
+
+const createObjectMap = (data) => {
+
```
Comment (firehudson) [inline buildParts.js:51]: Remove empty line

File: buildParts.js [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+		const keys = key.split('.');
+		keys.pop();
+
+		const updatedAmount = keys.length;
+		const updatedKey = keys[updatedAmount-1];
+		const updatedName = keys.join('.');
+		const checkName = key.slice(0, updatedName.length);
+		const valueName = key.slice(updatedName.length + 1, key.length);
+
+		if (updatedKey !== undefined) {
```
Comment (firehudson) [inline buildParts.js:72]: Replace `(updatedKey !== undefined)` with `(updatedKey)`

File: buildParts.js [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+			// And set it to the correct place
+			// TODO add check for old value of source/target
+			if (
+				inputObject.source?.type !== undefined
+				&& inputObject.source?.name !== undefined
+				&& inputObject.source?.value !== undefined
+				&& 'comparator' in inputObject
+				&& inputObject.target?.type !== undefined
+				&& inputObject.target?.value !== undefined
+			) {
```
Comment (firehudson) [inline buildParts.js:105]: Replace `if (
				inputObject.source?.type !== undefined
				&& inputObject.source?.name !== undefined
				&& inputObject.source?.value !== undefined
				&& 'comparator' in inputObject
				&& inputObject.target?.type !== undefined
				&& inputObject.target?.value !== undefined
			) {`

with 
`if (
				inputObject.source?.type
				&& inputObject.source?.name
				&& inputObject.source?.value
				&& 'comparator' in inputObject
				&& inputObject.target?.type
				&& inputObject.target?.value
			) {
`

File: buildParts.js [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+		}
+		// necessary for usage of old/new map
+		oldName = updatedName !== '' || updatedName !== undefined ? updatedName : oldName;
+	}
+
+	return topLevelMap;
+}
+
+function createObjectName(source) {
+
```
Comment (firehudson) [inline buildParts.js:131]: Remove empty line

File: buildParts.js [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+				objectName += `-${fixedValues[index++]}`;
+			}
+		}
+	})
+	objectName = objectName.replace(/ /g, '');
+	return {object, variableNames, objectName};
+}
+
+const buildFactPart = (source) => {
+	//
```
Comment (firehudson) [inline buildParts.js:187]: Remove this line

File: buildParts.js [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+	const sourceType = capitalizeFirstLetter(factName[factName.length-1]);
+	factName = `fact${collectionCapitalized}${sourceType}`;
+
+	// Check for variable usage in gql query
+	let factValue = sourceString.split('.')
+	factValue = factValue[factValue.length-1]
+
+	let fact;
+
+	if (variableNames.length === 0) {
```
Comment (firehudson) [inline buildParts.js:204]: Replace `if (variableNames.length === 0)` with `if (!variableNames.length)`

File: buildParts.js [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+		object = `()${functionCall}${object}`;
+		fact = `()${functionCall}${objectName}().${factValue}`;
+
+	} else {
+		const updatedVariableName = variableNames.join(',');
+		const updatedObject = object.replaceAll('$', '')
+
+		object = `(${updatedVariableName})${functionCall}${updatedObject}`;
+		fact = `(${updatedVariableName})${functionCall}${objectName}(${updatedVariableName}).${factValue}`;
+
```
Comment (firehudson) [inline buildParts.js:214]: Remove empty line

File: buildParts.js [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+	const resultMap = new Map();
+	resultMap.set('object', {[objectName]: object})
+	resultMap.set('fact', {[factName]: fact})
+	resultMap.set('variable', variableNames);
+
+	return resultMap;
+}
+
+const buildConditionPart = (inputMap, inputObject) => {
+
```
Comment (firehudson) [inline buildParts.js:226]: Remove empty line

File: buildParts.js [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+	const collection = objectName.split('By')[0];
+	const collectionCapitalized = capitalizeFirstLetter(collection);
+	const sourceType = capitalizeFirstLetter(source.name);
+
+	// Create function names - Condition
+	const formattedComparator = capitalizeFirstLetter(comparatorString);
+	const formattedTarget = capitalizeFirstLetter(removeQuotes(targetString));
+
+	let conditionName;
```
Comment (firehudson) [inline buildParts.js:249]: Move all variable declaration at top of file

File: buildParts.js [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+		condition = `(${updatedVariableName})${functionCall}${factName}(${updatedVariableName}) ${operatorString} ${targetString}`;
+	}
+
+	inputMap.set('condition', {[conditionName]: condition});
+
+	return inputMap;
+}
+
+const buildRulePart = (inputMap, ruleName) => {
+
```
Comment (firehudson) [inline buildParts.js:270]: Remove empty line

File: buildParts.js [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+		} else {
+			// get correct position from key
+			if (key.length !== oldKey.length) {
+				updatedKey = key.slice(oldKey.length + 1);
+			}
+
+			// add brackets and operator
+			// only if it's not the last key
+			if(key !== lastKey) {
+
```
Comment (firehudson) [inline buildParts.js:325]: Remove empty line

File: buildParts.js [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+	const brackets = closeBracket.repeat(counterBrackets);
+	ruleString += brackets;
+
+	const rule = `(${[...allUsedVariables].flat().join(',')})${functionCall}${ruleString}`;
+	resultMap.set('rule', {[updatedRuleName]: rule});
+
+	return resultMap;
+}
+
+const createFunction = (functionName, functionBody) => {
```
Comment (firehudson) [inline buildParts.js:356]: Convert function into single line arrow function

File: buildRule.js [hunk 21]
```diff
@@ -185,10 +132,21 @@ const build = (obj) => {
...
+  // objectMap.forEach((value, key) =>{
+  //   console.log(createFunction(key, value))
+  // })
+  // factMap.forEach((value, key) =>{
+  //   console.log(createFunction(key, value))
+  // })
+  // ruleMap.forEach((value, key) =>{
+  //   console.log(createFunction(key, value))
+  // })
+
```
Comment (firehudson) [inline buildRule.js:149]: Remove this code if not required

File: buildRule.js [hunk 22]
```diff
@@ -204,434 +162,20 @@ const buildCondition = (inputObject) => {
...
-    name: '${functionName}',
-    body: '${functionBody}'
-  })`
+  return ruleValue;
 }
 
-// buildRule(requestBody)
-build(requestBodyFact)
+build(requestBody);
+// build(requestBodyFact)
```
Comment (firehudson) [inline buildRule.js:179]: Remove commented code if not required

File: buildParts.js [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+	let oldName = Array.from(data.keys())[2].split('.');
+	oldName.pop();
+	oldName = oldName.join('.');
+
+	for (const [key, value] of data.entries()) {
+		const keys = key.split('.');
+		keys.pop();
+
+		const updatedAmount = keys.length;
+		const updatedKey = keys[updatedAmount-1];
```
Comment (firehudson) [inline buildParts.js:67]:
```suggestion
		const updatedKey = keys[updatedAmount - 1];
```

File: buildParts.js (hunk 12 above)
Comment (firehudson) [reply buildParts.js:131]: in JS conventionally we don't leave empty line at the start of function body

**Review by firehudson — CHANGES_REQUESTED** [review]

File: buildParts.js [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+  })`
+}
+
+module.exports = {
+	createObjectMap,
+	buildFactPart,
+	buildConditionPart,
+	buildRulePart,
+	createFunction
+}
```
Comment (firehudson) [inline buildParts.js:369]: Let's add EOF empty line

**Review by firehudson — CHANGES_REQUESTED** [review]

**Pull request merged**